### DISCUSSION TITLE
Deep.patch() returning new object different from original object

### DISCUSSION THREAD

**Discussion by st-clair-clarke** [rimbu-org/rimbu#139] [Q&A]
Hi
I have the following typescript domain models:

```
export interface IHistory {
	id: TId

	pc: IPc[]
}


export interface IHistoryState {
	id: TId

	domain: IHistory
}

```

My store update function is shown below:

```
	updatePC(complaints: IHistoryState['domain']['pc']) {
		store.update((state: IHistoryState) =>
			Deep.patch(state, (v) => [
				{
					domain: [{ pc: complaints }],
				},
			]),
		)
	}
```

The function updates the store, but the update object is not the same type as IHIstoryState. Below is a screenshot of the store:

![image](https://user-images.githubusercontent.com/3227590/216773758-528a31e8-2cce-4118-9b16-1e5d2379d0c8.png)

NB: 
1. Note that domain first child is an array, NOT and OBJECT as in IHistoryState
2. pc Object should be the first child of the updated object

I expected patch to return the object it was given, with the updated patched state; however, in my code it seems to be literally including the [] symbols that are mandatory when patch is being used.

It seems that something is incorrect in my updatePC function. How can I correct this issue?

Cheers

**Comment by vitoke** [comment]
Thanks for reaching out.

I created a [CodeSandbox](https://codesandbox.io/s/deep-patch-issue-qnjzx9?previewwindow=console&file=/src/index.ts) to try to recreate the issue you describe. Since I don't have the `store` you refer to, I created a function that directly uses `Deep.patch` inside and returns the result. Please check if I'm doing something different, but it seems to work fine there. Feel free to modify the sandbox and share the new link to show the issue you're having.

As a side note, are you sure you're on the latest version of the library? (currently 0.11.0)

**Comment by st-clair-clarke** [comment]
Thanks for your quick reply:
 "@rimbu/actor": "^0.11.1"
 "@rimbu/core": "^0.11.0",
 
 I am checking out the sandbox now.
 
 The actor package needs some documentation though.

I am using @ngneat/elf store (npm package)

**Comment by st-clair-clarke** [comment]
This [updated sandbox](https://codesandbox.io/s/deep-patch-issue-forked-lu5x58?file=/src/index.ts) is closer to what I am doing. The output and my RESULT is showing undefined though. Hope you can help.

**Comment by vitoke** [comment]
Thanks for updating the sandbox. I think the main issue I see there is that you are using a class for the `HistoryState`. The `Deep.patch` function is created to work on plain JS objects because it is very hard to clone class-based objects in a predictable way in JS. TypeScript also makes it really hard / impossible to warn about non-plain objects being supplied.

Your remark about `@rimbu/actor` (and reactor) is valid. It has mostly been in an exploratory stage so far, but I'm happy with the current version. Currently I have less time for this project but I hope that will change again soon. First I plan on adding support for serialisation to all the collections (work was already started), and then I will improve the documentation for the remaining packages.

**Comment by st-clair-clarke** [comment]
I use Typescript. The use of plain JS is minimized and not encouraged in the project. Have you considered using a class-transformer to bridge the gap, for example [class-transformer](https://www.npmjs.com/package/class-transformer)? Of course there are others. I will attempt to use it with the example and see what happens.

**Comment by vitoke** [comment]
In my projects I use typed plain objects. Instead of creating a class, you can create an interface or type for the expected plain objects. If you need validation, you can use a library like [zod](https://zod.dev/) that does validation and provides types at the same time. The issue is that a class instance cannot be reliably cloned, especially when it inherits from other classes.

**Comment by st-clair-clarke** [comment]
I will investigate your advice. Thanks.
